src/etos_test_runner/lib/log_area.py

```diff
@@ -0,0 +1,340 @@
+            "name": "",
+        }
+        published_url = upload["url"].format(**data)
+        self._artifact_published(artifact_created, published_url)
+
+    def __upload(self, context, log, name, folder):
+        """Upload log to a storage location.
+
+        :param context: Context for the http request.
+        :type context: str
+        :param log: Path to the log to upload.
+        :type log: str
+        :param name: Name of file to upload.
+        :type name: str
+        :param folder: Folder to upload to.
+        :type folder: str
+        :return: URI where log was uploaded to.
+        :rtype: str
+        """
+        upload = deepcopy(self.log_area.get("upload"))
+        data = {"context": context, "name": name, "folder": folder}
+
+        # ETOS Library, for some reason, uses the key 'verb' instead of 'method'
+        # for HTTP method.
+        upload["verb"] = upload.pop("method")
+        upload["url"] = upload["url"].format(**data)
+        upload["timeout"] = upload.get("timeout", 30)
+        if upload.get("auth"):
+            upload["auth"] = self.__auth(**upload["auth"])
+
+        with open(log, "rb") as log_file:
+            for _ in range(3):
+                request_generator = self.__retry_upload(log_file=log_file, **upload)
+                try:
+                    for response in request_generator:
+                        self.logger.debug("%r", response)
+                        if not upload.get("as_json", True):
+                            self.logger.debug("%r", response.text)
+                        self.logger.info("Uploaded log %r.", log)
+                        self.logger.info("Upload URI          %r", upload["url"])
+                        self.logger.info("Data:               %r", data)
+                        break
+                    break
+                except:  # noqa pylint:disable=bare-except
+                    self.logger.error("%r", traceback.format_exc())
+                    self.logger.error("Failed to upload log!")
+                    self.logger.error("Attempted upload of %r", log)
+        return upload["url"]
+
+    def __retry_upload(
+        self, verb, url, log_file, timeout=None, as_json=True, **requests_kwargs
+    ):  # pylint:disable=too-many-arguments
+        """Attempt to connect to url for x time.
+
+        :param verb: Which HTTP verb to use. GET, PUT, POST
+                     (DELETE omitted)
+        :type verb: str
+        :param url: URL to retry upload request
+        :type url: str
+        :param log_file: Opened log file to upload.
+        :type log_file: file
+        :param timeout: How long, in seconds, to retry request.
+        :type timeout: int or None
+        :param as_json: Whether or not to return json instead of response.
+        :type as_json: bool
+        :param request_kwargs: Keyword arguments for the requests command.
+        :type request_kwargs: dict
+        :return: HTTP response or json.
```

Comment:
remove ?

Reply:
This one actually returns

Reply:
Ah true, perhaps an update to the return type?

Reply:
Nope. It's correct.